### ISSUE TITLE
Duck typing not working inside assertion

### ISSUE THREAD

**Issue by rafaferry** [sylvainpolletvillard/ObjectModel#78]
Hello,

I'm not sure if this is a bug report or just a question. My assumption is that in the following test case should work, but it throws an error.
The line "**variavel.inicioVigencia.isAnterior(variavel.terminoVigencia);**" doens't work because inicioVigencia should be a instance of Competencia, but it is not, so there is no "isAnterior" method.

```
const ObjectModel = require('objectmodel').ObjectModel;
const moment = require('moment');

it('should duck type inside assert', () => {
  class Competencia extends ObjectModel({
    ano: Number,
    mes: Number,
  }).assert(comp => comp.mes > 0 && comp.mes <= 12, 'mes.invalido') {
    isAnterior(competencia) {
      const myMoment = moment(`${this.mes}/${this.ano}`, 'MM/YYYY');
      const otherMoment = moment(`${competencia.mes}/${competencia.ano}`, 'MM/YYYY');

      return myMoment.isBefore(otherMoment);
    }
  }

  const Variavel = ObjectModel({
    inicioVigencia: Competencia,
    terminoVigencia: [Competencia],
  }).assert((variavel) => {
    if (variavel.terminoVigencia == null) {
      return true;
    }
    return variavel.inicioVigencia.isAnterior(variavel.terminoVigencia);
  }, 'vigencia.invalida');

  const v = new Variavel({
    inicioVigencia: { mes: 1, ano: 2017 },
    terminoVigencia: { mes: 2, ano: 2017 },
    ordem: 1,
  });

  expect(v).to.exist();
});
```

**Comment by sylvainpolletvillard** [comment]
Hi,

I confirm it's a bug, and I think I have a fix. Thanks for the report

**Comment by sylvainpolletvillard** [comment]
Hmm, this may be more complicated than what I thought. I made a simpler example:

```js
const A = ObjectModel({ foo: String })
A.prototype.getFoo = function(){ return this.foo }

const B = ObjectModel({ a: A }).assert(b => b.a.getFoo() === "bar")

const b = B({ a: { foo: "bar" } }) // error, getFoo is not a function
```

The problem is, you can't duck-type and validate at the same time :
- `b.a` will be casted to A once `b` has been validated as B
- `b` will be validated as B if its assertion return `true`, but that implies that `b.a` has been casted to A

So there is a logical conflict. It would be possible if we were doing the type-checking in 2 distinct steps:
1. Check definition and duck-type everything you can
2. Check the assertions

But the validation has to be one single step, because we can't stop halfway. Otherwise, if the definition is OK but one assertion failed, we could have duck-typed and mutated some properties even if the object is finally declared invalid. This is something we want to avoid.

**To put it simply, an assertion should not assume that a property has been duck-typed ; because the property will be duck-typed only if the model instance is validated, and therefore if the assertion passed**

**Comment by sylvainpolletvillard** [comment]
Ok, I close the issue then.
